pytorch_lightning/callbacks/timer.py

```diff
@@ -0,0 +1,133 @@
+            If ``interval`` is not one of the supported choices.
+
+    Example::
+        from pytorch_lightning import Trainer
+        from pytorch_lightning.callbacks import Timer
+
+        # stop training after 12 hours
+        timer = Timer(duration="00:12:00:00")
+
+        # or provide a datetime.timedelta
+        from datetime import timedelta
+        timer = Timer(duration=timedelta(weeks=1))
+
+        # or provide a dictionary
+        timer = Timer(duration=dict(weeks=4, days=2))
+
+        # force training to stop after given time limit
+        trainer = Trainer(callbacks=[timer])
+    """
+
+    def __init__(
+        self,
+        duration: Union[str, timedelta, Dict[str, int]],
+        interval: str = Interval.step,
+        verbose: bool = True,
+    ):
+        super().__init__()
+        if isinstance(duration, str):
+            dhms = duration.strip().split(":")
+            dhms = [int(i) for i in dhms]
+            duration = timedelta(days=dhms[0], hours=dhms[1], minutes=dhms[2], seconds=dhms[3])
+        if isinstance(duration, dict):
+            duration = timedelta(**duration)
+        if interval not in set(Interval):
+            raise MisconfigurationException(
+                f"Unsupported parameter value `Timer(interval={interval})`. Possible choices are:"
+                f" {', '.join(set(Interval))}"
+            )
+        self._duration = duration
+        self._interval = interval
+        self._verbose = verbose
+        self._start_time = None
+        self._offset = timedelta()
+
+    @property
+    def start_time(self) -> datetime:
+        return self._start_time
+
+    @property
+    def time_elapsed(self) -> timedelta:
+        if self._start_time is None:
+            return self._offset
+        return datetime.now() - self._start_time + self._offset
+
+    @property
+    def time_remaining(self) -> timedelta:
+        return self._duration - self.time_elapsed
+
+    def on_train_start(self, trainer, *args, **kwargs) -> None:
+        self._start_time = datetime.now()
+
+    def on_train_batch_end(self, trainer, *args, **kwargs) -> None:
+        if self._interval != Interval.step:
+            return
+        self._check_time_remaining(trainer)
+
+    def on_train_epoch_end(self, trainer, *args, **kwargs) -> None:
+        if self._interval != Interval.epoch:
+            return
+        self._check_time_remaining(trainer)
+
+    def on_save_checkpoint(self, trainer, pl_module, checkpoint: Dict[str, Any]) -> Dict[str, Any]:
+        return {
+            "time_elapsed": self.time_elapsed,
+        }
+
+    def on_load_checkpoint(self, callback_state: Dict[str, Any]):
+        self._offset = callback_state.get("time_elapsed", timedelta())
+
+    def _check_time_remaining(self, trainer) -> None:
+        should_stop = self.time_elapsed >= self._duration
+        should_stop = trainer.accelerator.broadcast(should_stop)
+        trainer.should_stop = trainer.should_stop or should_stop
```

Comment:
n00b question: why the or?

Reply:
I don't want to overwrite the decision by trainer. Only when the early stopping triggers `should_stop = True` should we be allowed to write the trainer attribute, but not when `should_stop = False` and `trainer.should_stop = True`